src/solvers/refinement/string_refinement.cpp

```diff
@@ -1011,31 +1075,71 @@ bool string_refinementt::check_axioms()
     }
   }
 
-  // tells whether one of the not_contains constraint can be violated
-  bool violated_not_contains=false;
+  // Maps from indexes of violated not_contains axiom to a witness of violation
+  std::map<size_t, exprt> violated_not;
```

Comment:
also maybe keep the `violated_not_contains` name, as that could be understood as meaning "not violated"